src/lib/route-builder.js

```diff
@@ -0,0 +1,165 @@
+'use strict'
+const _ = require('lodash')
+const uuid = require('uuid4')
+const BigNumber = require('bignumber.js')
+const AssetsNotTradedError = require('../errors/assets-not-traded-error')
+const UnacceptableAmountError = require('../errors/unacceptable-amount-error')
+
+class RouteBuilder {
+  /**
+   * @param {RoutingTables} routingTables
+   * @param {PrecisionCache} precisionCache
+   * @param {Object} config
+   * @param {Integer} config.minMessageWindow seconds
+   * @param {Number} config.slippage
+   * @param {Object} config.ledgerCredentials
+   */
+  constructor (routingTables, precisionCache, config) {
+    this.baseURI = routingTables.baseURI
+    this.routingTables = routingTables
+    this.precisionCache = precisionCache
+    this.minMessageWindow = config.minMessageWindow
+    this.slippage = config.slippage
+    this.ledgerCredentials = config.ledgerCredentials
+  }
+
+  /**
+   * @param {Object} query
+   * @param {String} query.sourceLedger
+   * @param {String} query.sourceAccount
+   * @param {String} query.sourceAmount
+   * @param {String} query.destinationLedger
+   * @param {String} query.destinationAccount
+   * @param {String} query.destinationAmount
+   * @returns {Transfer}
+   */
+  * getQuote (query) {
+    const _nextHop = this._findNextHop(query)
+    if (!_nextHop) throwAssetsNotTradedError()
+    if (query.sourceAmount) {
+      _nextHop.finalAmount = (new BigNumber(_nextHop.finalAmount)).times(1 - this.slippage).toString()
+    } else { // fixed destinationAmount
+      _nextHop.sourceAmount = (new BigNumber(_nextHop.sourceAmount)).times(1 + this.slippage).toString()
+    }
+    const nextHop = yield this._roundHop(_nextHop)
+
+    return {
+      ledger: nextHop.sourceLedger,
+      debits: [{ account: query.sourceAccount, amount: nextHop.sourceAmount }],
+      credits: [{
+        account: this.ledgerCredentials[nextHop.sourceLedger].account_uri,
+        amount: nextHop.sourceAmount,
+        memo: {
+          destination_transfer: {
+            ledger: nextHop.finalLedger,
+            debits: [{ account: null, amount: nextHop.finalAmount }],
+            credits: [{ account: query.destinationAccount, amount: nextHop.finalAmount }]
+          }
+        }
+      }],
+      _hop: nextHop
+    }
+  }
+
+  /**
+   * Given a source transfer with an embedded final transfer, get the next
+   * transfer in the chain.
```

Comment:
Not strictly necessary for this PR but it would be good to have some docs that explains in detail how this works